### PR TITLE
Refactor trait set container

### PR DESCRIPTION
**Pull request by midhun-pm**
This PR refactors the new `TraitSet` implementation to try and match the refactored `TraitList` implementation. 

Some unresolved comments in PR #922 have been addressed here.

### CODE REVIEW

File: traits/trait_set_object.py [hunk 1]
```diff
@@ -338,179 +300,93 @@ def symmetric_difference_update(self, value):
             self.notify(removed, added)
```
Comment (kitchoi) [inline traits/trait_set_object.py:300]: Comment on lines 298-299 (GitHub does not allow me comment directly on them).
Nit: If you replace these two lines with a call to `super().symmetric_difference_update` above the `if removed or added:`, that would also be fine and would be more consistent with the other changes, something like that:
```
        super().symmetric_difference_update(removed | added)
        if removed or added:
            self.notify(removed, added)
```

File: traits/trait_set_object.py [hunk 2]
```diff
@@ -38,31 +37,48 @@ def __init__(self, removed=None, added=None):
...
+        callable should accept a single item from the list and return
+        the transformed item, raising TraitError for invalid items. If
+        not given, no item validation is performed.
     notifiers : list of callable
         A list of callables with the signature::
 
             notifier(removed, added)
 
+        If this argument is not given, the list of notifiers is initially
+        empty.
```
Comment (kitchoi) [inline traits/trait_set_object.py:74]: I see you have removed the notes on what `removed` and `added` are in individual methods.
Are you going to mention what `removed` and `added` are here or is the description added somewhere?

File: traits/trait_set_object.py [hunk 3]
```diff
@@ -38,31 +37,48 @@ def __init__(self, removed=None, added=None):
...
-        The value for the set
-    validator : callable
-        Called to validate items in the set
+    value : iterable
+        Iterable providing the items for the set.
+    item_validator : callable, optional
+        Called to validate and/or transform items added to the list. The
+        callable should accept a single item from the list and return
+        the transformed item, raising TraitError for invalid items. If
+        not given, no item validation is performed.
```
Comment (kitchoi) [inline traits/trait_set_object.py:67]: "list" -> "set"?

File: traits/trait_set_object.py [hunk 4]
```diff
@@ -38,31 +37,48 @@ def __init__(self, removed=None, added=None):
...
+        self.removed = removed
+
         if added is None:
             added = set()
-
-        self.removed = removed
         self.added = added
 
 
+# Default item validator for TraitList.
```
Comment (kitchoi) [inline traits/trait_set_object.py:47]: "TraitList" -> "TraitSet"?

File: traits/trait_set_object.py (hunk 2 above)
Comment (midhun-pm) [reply traits/trait_set_object.py:77]: I just added a description here. Should be enough to understand what they are.

File: traits/trait_set_object.py [hunk 5]
```diff
@@ -71,55 +87,20 @@ class TraitSet(set):
...
-        value : set
-            The new items being added to the set.
-
-        Returns
-        -------
-        values : set
-            The set of validated values.
+        # We need a __new__ in addition to __init__ in order to properly
+        # support unpickling: the 'append' or 'extend' methods may be
+        # called during unpickling, triggering item validation.
```
[the remark this answers is not in the document]
Comment (midhun-pm) [reply traits/trait_set_object.py]: Need to test if __new__ is needed. I think we'll just keep it for consistency with `TraitList`. Will remove the comment though.

File: traits/tests/test_trait_set_object.py [hunk 6]
```diff
@@ -13,22 +13,19 @@
...
 from traits.api import HasTraits, Set, Str
 from traits.trait_errors import TraitError
-from traits.trait_set_object import TraitSet
+from traits.trait_set_object import TraitSet, accept_anything
 from traits.trait_types import _validate_int
 
 
 def int_validator(value):
-    return {_validate_int(v) for v in value}
+    return _validate_int((value))
```
[the remark this answers is not in the document]
Comment (midhun-pm) [reply traits/tests/test_trait_set_object.py:21]: Fixed it here.

File: traits/tests/test_trait_set_object.py [hunk 7]
```diff
@@ -354,13 +351,13 @@ def validator(added):
         ts ^= set(["2", 3, 4])
 
         # then
-        self.assertEqual(validator_args, set([3, 4]))
+        self.assertEqual(validator_args, set([1, 2, 3, 4]))
```
[the remark this answers is not in the document]
Comment (midhun-pm) [reply traits/tests/test_trait_set_object.py]: Thanks!

File: traits/trait_set_object.py (hunk 1 above)
Comment (midhun-pm) [reply traits/trait_set_object.py:302]: Thanks!

**Comment by midhun-pm** [comment]
@kitchoi I've addressed the comments, can I merge this?

File: traits/trait_set_object.py [hunk 8]
```diff
@@ -38,88 +37,73 @@ def __init__(self, removed=None, added=None):
...
     Parameters
     ----------
-    value : set
-        The value for the set
-    validator : callable
-        Called to validate items in the set
+    value : iterable
+        Iterable providing the items for the set.
+    item_validator : callable, optional
+        Called to validate and/or transform items added to the list. The
```
Comment (kitchoi) [inline traits/trait_set_object.py:64]: There is one more "list" here that needs to be changed to "set".

File: traits/trait_set_object.py [hunk 9]
```diff
@@ -38,88 +37,73 @@ def __init__(self, removed=None, added=None):
...
+        callable should accept a single item from the set and return
+        the transformed item, raising TraitError for invalid items. If
+        not given, no item validation is performed.
     notifiers : list of callable
         A list of callables with the signature::
 
             notifier(removed, added)
 
+        Where 'added' is a set containing the new values that will be added.
+        And 'removed' is a set containing the old values that will be removed.
```
Comment (kitchoi) [inline traits/trait_set_object.py:74]: Nit: When the notifier is called, the change has happened. So the future tense may be better changed to either present or present perfect tense? I am also no native speaker to be relied on for such things... :(

**Pull request merged**